packages-exp/auth-exp/src/model/action_code_url.ts

```diff
@@ -0,0 +1,102 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { AuthErrorCode, AUTH_ERROR_FACTORY } from '../core/errors';
+import { Operation } from './action_code_info';
+import { Auth } from './auth';
+
+/**
+ * Enums for fields in URL query string.
+ * @enum {string}
+ */
+enum QueryField {
+  API_KEY = 'apiKey',
+  CODE = 'oobCode',
+  CONTINUE_URL = 'continueUrl',
+  LANGUAGE_CODE = 'languageCode',
+  MODE = 'mode',
+  TENANT_ID = 'tenantId'
+}
+
+/**
+ * Map from mode string in action code URL to Action Code Info operation.
+ */
+const ModeToOperationMap: { [key: string]: Operation } = {
+  'recoverEmail': Operation.RECOVER_EMAIL,
+  'resetPassword': Operation.PASSWORD_RESET,
+  'signIn': Operation.EMAIL_SIGNIN,
+  'verifyEmail': Operation.VERIFY_EMAIL,
+  'verifyAndChangeEmail': Operation.VERIFY_AND_CHANGE_EMAIL,
+  'revertSecondFactorAddition': Operation.REVERT_SECOND_FACTOR_ADDITION
+};
+
+/**
+ * Maps the mode string in action code URL to Action Code Info operation.
+ */
+function parseMode(mode: string | null): Operation | null {
+  return mode ? ModeToOperationMap[mode] || null : null;
+}
+
+function parseDeepLink(url: string): string {
+  const uri = new URL(url);
+  const link = uri.searchParams.get('link');
+  // Double link case (automatic redirect).
+  const doubleDeepLink = link ? new URL(link).searchParams.get('link') : null;
+  // iOS custom scheme links.
+  const iOSDeepLink = uri.searchParams.get('deep_link_id');
+  const iOSDoubleDeepLink = iOSDeepLink
+    ? new URL(iOSDeepLink).searchParams.get('link')
+    : null;
+  return iOSDoubleDeepLink || iOSDeepLink || doubleDeepLink || link || url;
+}
+
+export class ActionCodeURL {
```

Comment:
@samhorlbeck should I split this into an interface & implementation like we have with user & auth?

Reply:
Whoops yeah this whole thing should be moved to core. I don't think you'll need an interface for it (we need that for user and auth to avoid circular dependency issues... I don't anticipate that being a problem for this module)